# **KPMG VIRTUAL INTERNSHIP PROJECT**
# TASK: 1 - Data Quality Assessment
Assessment of data quality and completeness in preparation for analysis. The client (Sprocket Central Pty Ltd) has provided KPMG with 3 datasets:

1.Customer Demographic

2.Customer Addresses

3.Transactions data in the past 3 months

# Data Quality Framework Table
Using the dimensions included in the Data Quality Framework, I will assess the quality of these datasets. Followings are the dimesnions provided by the Data Quality Framework:
- Completeness : How much information all entities have. Number of missing values.
- Consistency : How conistent is your Data. Number of inconsistencies in your data.
- Accuarcy : How accurate is your Data. Number of errors in you data.
- Relevancy/Auditability : Relevanct data in your entities. Number of irrelavant values.
- Validity : Validated data with allowable values.
- Uniqueness: How much uniques is your data. Number of duplicated values.
- Timeliness: Updated data. Current data.

# **1) IMPORTING THE REQUIRED LIBRARIES**

In [2]:
import pandas as pd

# **2) READING THE COMPLETE DATASET**

In [3]:
data = pd.ExcelFile("KPMG_VI_New_raw_data_update_final.xlsx")

# **3) READ EACH SEPERATE FILES IN THE DATASET**

In [4]:
Transactions = pd.read_excel(data, 'Transactions')
NewCustomerList = pd.read_excel(data, 'NewCustomerList')
CustomerDemographic = pd.read_excel(data, 'CustomerDemographic')
CustomerAddress = pd.read_excel(data, 'CustomerAddress')

<ipython-input-4-7e31c9dbc7f2>:2: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  NewCustomerList = pd.read_excel(data, 'NewCustomerList')
<ipython-input-4-7e31c9dbc7f2>:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  CustomerDemographic = pd.read_excel(data, 'CustomerDemographic')


# **4) EXPLORING THE TRANSACTIONS FILE**

In [5]:
data

In [6]:
Transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

## Checking Consistency and Validity of Dataset

In [7]:
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

## Highlights of Consistency and Validity in Transactions
Transactions dataset has 20000 records with 13 columns.
- Out of which, 3 are of datatype **int64** which are keys.
- One is the date **datetime64** in format **MM/DD/YYYY** (check the dataset). The date format used to capture DOB of customers is **YYYY-MM-DD**. It would be better if it is kept consistent.
- Another one is Online Order which is captured in a column of **float64** datatype, however the values are **boolean**, that is true and false.
- 5 columns are of datatype **object** which are order_status, brand, product_line, product_class, product_size.
- Last 3 columns are of datatype **float64** again from which one of them is a date (i.e., product_first_sold_date) and should be in the standard **datetime64** format.

In [8]:
Transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [9]:
#Checking the shape of the data
Transactions.shape

(20000, 13)

## Checking Completeness of the Dataset

In [10]:
#Checking for null values
Transactions.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

There are missing values in 7 columns.

In [11]:
#Checking for null values
total_null_values = Transactions.isnull().sum()
# calculating total values
total_values = Transactions.count().sort_values(ascending=True)

# calculating the percentage of null values
null_values_percentage = total_null_values/total_values *100

# converting to dataframe of missing values
missing_values = pd.concat({'Total Values' : total_values, 'Null_values': total_null_values, 'Percentage of Missing Values': null_values_percentage}, axis=1)

# display missing values
print(missing_values)

                         Total Values  Null_values  \
online_order                    19640          360   
brand                           19803          197   
product_line                    19803          197   
product_class                   19803          197   
product_size                    19803          197   
standard_cost                   19803          197   
product_first_sold_date         19803          197   
transaction_id                  20000            0   
product_id                      20000            0   
customer_id                     20000            0   
transaction_date                20000            0   
order_status                    20000            0   
list_price                      20000            0   

                         Percentage of Missing Values  
online_order                                 1.832994  
brand                                        0.994799  
product_line                                 0.994799  
product_class      

## Highlights of Completeness in Transactions
- Order Online columns has about 1.83% of null values. There are 360 records in which order_online was not captured.
- Columns brand, product_line, product_class, product_size, standard_cost, product_first_sold_date also has a percentage of 0.995% missing values that is 197 null values, which should not be missing if product_id is inherited and the details of the product cannot be missing.

## Checking Accuracy of Dataset

In [12]:
# checking a single product id and its details
bool_series = Transactions['product_id'] == 0

product_id_0 = Transactions[bool_series]

#view the product details
print(product_id_0[['brand', 'product_line','product_class']])

                brand product_line product_class
34     Norco Bicycles         Road        medium
39     Norco Bicycles         Road        medium
54     Norco Bicycles     Standard           low
60         OHM Cycles         Road          high
63      Trek Bicycles     Standard        medium
...               ...          ...           ...
19921  Norco Bicycles         Road        medium
19941      OHM Cycles     Standard           low
19967        WeareA2B     Standard        medium
19987  Norco Bicycles         Road        medium
19988  Norco Bicycles     Standard           low

[1378 rows x 3 columns]


## Highlights of Accuracy in Transactions
A single product ID should be referencing a single product with unique values.

## Checking for duplicate values

In [13]:
Transactions.duplicated().sum()

0

There are no duplicate values, so the data is unique

In [14]:
#check for uniqueness of each column
Transactions.nunique()

transaction_id             20000
product_id                   101
customer_id                 3494
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   296
standard_cost                103
product_first_sold_date      100
dtype: int64

# **4.1) Exploring the columns of Transactions**

In [15]:
Transactions.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [16]:
Transactions['order_status'].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [17]:
Transactions['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [18]:
Transactions['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [19]:
Transactions['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [20]:
Transactions['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

In [21]:
Transactions['product_first_sold_date']

0        41245.0
1        41701.0
2        36361.0
3        36145.0
4        42226.0
          ...   
19995    37823.0
19996    35560.0
19997    40410.0
19998    38216.0
19999    36334.0
Name: product_first_sold_date, Length: 20000, dtype: float64

In [22]:
#convert date column from integer to datetime
Transactions['product_first_sold_date'] = pd.to_datetime(Transactions['product_first_sold_date'], unit='s')
Transactions['product_first_sold_date'].head(20)

0    1970-01-01 11:27:25
1    1970-01-01 11:35:01
2    1970-01-01 10:06:01
3    1970-01-01 10:02:25
4    1970-01-01 11:43:46
5    1970-01-01 10:50:31
6    1970-01-01 09:29:25
7    1970-01-01 11:05:15
8    1970-01-01 09:17:35
9    1970-01-01 10:36:56
10   1970-01-01 11:19:44
11   1970-01-01 11:42:52
12   1970-01-01 09:35:27
13   1970-01-01 09:36:26
14   1970-01-01 10:36:33
15   1970-01-01 10:31:13
16   1970-01-01 10:36:46
17   1970-01-01 09:24:48
18   1970-01-01 11:05:15
19   1970-01-01 10:22:17
Name: product_first_sold_date, dtype: datetime64[ns]

In [23]:
Transactions['product_first_sold_date']

0       1970-01-01 11:27:25
1       1970-01-01 11:35:01
2       1970-01-01 10:06:01
3       1970-01-01 10:02:25
4       1970-01-01 11:43:46
                ...        
19995   1970-01-01 10:30:23
19996   1970-01-01 09:52:40
19997   1970-01-01 11:13:30
19998   1970-01-01 10:36:56
19999   1970-01-01 10:05:34
Name: product_first_sold_date, Length: 20000, dtype: datetime64[ns]

The values in the product_first_sold_date columns are not correct as it shows everything happening the same day at different times.

# **5) EXPLORING NEW CUSTOMER LIST DATA SET**

In [24]:
NewCustomerList

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0      Chickie    Brister    Male                                   86   
1        Morly     Genery    Male                                   69   
2      Ardelis  Forrester  Female                                   10   
3       Lucine      Stutt  Female                                   64   
4      Melinda     Hadlee  Female                                   34   
..         ...        ...     ...                                  ...   
995  Ferdinand  Romanetti    Male                                   60   
996       Burk    Wortley    Male                                   22   
997   Melloney      Temby  Female                                   17   
998     Dickie    Cubbini    Male                                   30   
999      Sylas    Duffill    Male                                   56   

           DOB                     job_title job_industry_category  \
0   1957-07-12               General Manager         Manufacturing   
1   1970-03-22           Structural Engineer              Property   
2   1974-08-28        Senior Cost Accountant    Financial Services   
3   1979-01-28    Account Representative III         Manufacturing   
4   1965-09-21             Financial Analyst    Financial Services   
..         ...                           ...                   ...   
995 1959-10-07                     Paralegal    Financial Services   
996 2001-10-17        Senior Sales Associate                Health   
997 1954-10-05  Budget/Accounting Analyst IV    Financial Services   
998 1952-12-17             Financial Advisor    Financial Services   
999 1955-10-02           Staff Accountant IV              Property   

        wealth_segment deceased_indicator owns_car  ...  state    country  \
0        Mass Customer                  N      Yes  ...    QLD  Australia   
1        Mass Customer                  N       No  ...    NSW  Australia   
2    Affluent Customer                  N       No  ...    VIC  Australia   
3    Affluent Customer                  N      Yes  ...    QLD  Australia   
4    Affluent Customer                  N       No  ...    NSW  Australia   
..                 ...                ...      ...  ...    ...        ...   
995  Affluent Customer                  N       No  ...    NSW  Australia   
996      Mass Customer                  N       No  ...    NSW  Australia   
997  Affluent Customer                  N      Yes  ...    QLD  Australia   
998      Mass Customer                  N      Yes  ...    QLD  Australia   
999      Mass Customer                  N      Yes  ...    NSW  Australia   

     property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                     6        0.53      0.6625     0.828125     0.703906   
1                    11        0.46      0.4600     0.575000     0.488750   
2                     5        0.84      0.8400     0.840000     0.840000   
3                     1        0.45      0.5625     0.562500     0.562500   
4                     9        0.52      0.5200     0.650000     0.650000   
..                  ...         ...         ...          ...          ...   
995                   7        0.83      0.8300     0.830000     0.830000   
996                  10        0.59      0.5900     0.737500     0.626875   
997                   2        1.00      1.2500     1.250000     1.250000   
998                   2        0.80      1.0000     1.000000     0.850000   
999                   9        0.46      0.5750     0.718750     0.610938   

     Unnamed: 20  Rank     Value  
0              1     1  1.718750  
1              1     1  1.718750  
2              1     1  1.718750  
3              4     4  1.703125  
4              4     4  1.703125  
..           ...   ...       ...  
995          996   996  0.374000  
996          997   997  0.357000  
997          997   997  0.357000  
998          997   997  0.357000  
999         1000  1000  0.340000  

[10

In [25]:
NewCustomerList.head(5)

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.53      0.6625     0.828125     0.703906   
1                  11        0.46      0.4600     0.575000     0.488750   
2                   5        0.84      0.8400     0.840000     0.840000   
3                   1        0.45      0.5625     0.562500     0.562500   
4                   9        0.52      0.5200     0.650000     0.650000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [26]:
NewCustomerList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [27]:
NewCustomerList.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean             7.397000     0.745860     0.838847     0.943651     0.870802   
std              2.758804     0.203589     0.252064     0.301854     0.286625   
min              1.000000     0.400000     0.400000     0.400000     0.340000   
25%              6.000000     0.570000     0.625000     0.712500     0.648125   
50%              8.000000     0.740000     0.830000     0.912500     0.840750   
75%              9.000000     0.922500     1.025000     1.162500     1.062500   
max             12.000000     1.100000     1.375000     1.718750     1.718750   

       Unnamed: 20         Rank        Value  
count  1000.000000  1000.000000  1000.000000  
mean    498.819000   498.819000     0.881714  
std     288.810997   288.810997     0.293525  
min       1.000000     1.000000     0.340000  
25%     250.000000   250.000000     0.649531  
50%     500.000000   500.000000     0.860000  
75%     750.250000   750.250000     1.075000  
max    1000.000000  1000.000000     1.718750

In [28]:
#Dropping the unnamed columns
NewCustomerList.drop(['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20'], axis=1, inplace=True)

In [29]:
#Checking the shape of the dataset
NewCustomerList.shape

(1000, 18)

In [30]:
#Checking for null values
NewCustomerList.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

There are missing values in 4 columns.

## Checking Completeness of Datasets

In [72]:
# looking for the null values
total_null_values = NewCustomerList.isnull().sum()

# calculating total values
total_values = NewCustomerList.count().sort_values(ascending=True)

# calculating the percentage of null values
null_values_percentage = total_null_values/total_values *100

# converting to dataframe of missing values
missing_values_NewCustomerList = pd.concat({'Total Values' : total_values, 'Null_values': total_null_values, 'Percentage of Missing Values': null_values_percentage}, axis=1)

# display missing values
print(missing_values_NewCustomerList)

                                     Total Values  Null_values  \
job_industry_category                         835          165   
job_title                                     894          106   
last_name                                     971           29   
DOB                                           983           17   
first_name                                   1000            0   
property_valuation                           1000            0   
country                                      1000            0   
state                                        1000            0   
postcode                                     1000            0   
address                                      1000            0   
deceased_indicator                           1000            0   
owns_car                                     1000            0   
Rank                                         1000            0   
wealth_segment                               1000            0   
past_3_yea

In [31]:
#Checking for duplicate values
NewCustomerList.duplicated().sum()

0

There are no duplicate values.

In [32]:
#Checking for uniquess of each column
NewCustomerList.nunique()

first_name                              940
last_name                               961
gender                                    3
past_3_years_bike_related_purchases     100
DOB                                     958
job_title                               184
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        1
owns_car                                  2
tenure                                   23
address                                1000
postcode                                522
state                                     3
country                                   1
property_valuation                       12
Rank                                    324
Value                                   324
dtype: int64

# **5.1) Exploring the columns of New Customer Dataset**

In [33]:
NewCustomerList.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Rank', 'Value'],
      dtype='object')

In [34]:
NewCustomerList['gender'].value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [35]:
NewCustomerList[NewCustomerList.gender == "U"]

first_name    last_name gender  past_3_years_bike_related_purchases DOB  \
59       Normy     Goodinge      U                                    5 NaT   
226      Hatti     Carletti      U                                   35 NaT   
324   Rozamond       Turtle      U                                   69 NaT   
358      Tamas      Swatman      U                                   65 NaT   
360      Tracy   Andrejevic      U                                   71 NaT   
374     Agneta      McAmish      U                                   66 NaT   
434      Gregg      Aimeric      U                                   52 NaT   
439      Johna       Bunker      U                                   93 NaT   
574    Harlene         Nono      U                                   69 NaT   
598   Gerianne       Kaysor      U                                   15 NaT   
664     Chicky      Sinclar      U                                   43 NaT   
751    Adriana  Saundercock      U                                   20 NaT   
775     Dmitri        Viant      U                                   62 NaT   
835      Porty       Hansed      U                                   88 NaT   
883      Shara     Bramhill      U                                   24 NaT   
904       Roth         Crum      U                                    0 NaT   
984    Pauline     Dallosso      U                                   82 NaT   

                        job_title job_industry_category     wealth_segment  \
59            Associate Professor                    IT      Mass Customer   
226               Legal Assistant                    IT  Affluent Customer   
324               Legal Assistant                    IT      Mass Customer   
358       Assistant Media Planner         Entertainment  Affluent Customer   
360                 Programmer II                    IT      Mass Customer   
374  Structural Analysis Engineer                    IT      Mass Customer   
434              Internal Auditor                    IT      Mass Customer   
439                Tax Accountant                    IT      Mass Customer   
574       Human Resources Manager                    IT      Mass Customer   
598               Project Manager                    IT  Affluent Customer   
664                      Operator                    IT     High Net Worth   
751                         Nurse                    IT     High Net Worth   
775                     Paralegal    Financial Services  Affluent Customer   
835               General Manager                    IT      Mass Customer   
883                           NaN                    IT  Affluent Customer   
904               Legal Assistant                    IT      Mass Customer   
984    Desktop Support Technician                    IT  Affluent Customer   

    deceased_indicator owns_car  tenure                  address  postcode  \
59                   N       No       4      7232 Fulton Parkway      3810   
226                  N      Yes      11            6 Iowa Center      2519   
324                  N      Yes       3  57025 New Castle Street      3850   
358                  N       No       5       78 Clarendon Drive      4551   
360                  N      Yes      11  5675 Burning Wood Trail      3030   
374                  N       No      15           5773 Acker Way      4207   
434                  N       No       7      72423 Surrey Street      3753   
439                  N      Yes      14         3686 Waubesa Way      3065   
574                  N       No      12  0307 Namekagon Crossing      2170   
598                  N       No       5           882 Toban Lane      2121   
664                  N      Yes       0        5 Red Cloud Place      3222   
751                  N      Yes      14         82 Gina Junction      3806   
775                  N       No       5     95960 Warner Parkway      3842   
835                  N       No      13     768 Southridge Drive      2112   
883   

There are 17 rows with unknown/unspecified gender.

In [36]:
NewCustomerList['DOB'].value_counts()

1998-02-05    2
1978-01-15    2
1977-11-08    2
1951-11-28    2
1979-07-28    2
             ..
1945-08-08    1
1943-08-27    1
1999-10-24    1
1976-01-24    1
1955-10-02    1
Name: DOB, Length: 958, dtype: int64

In [37]:
NewCustomerList['job_industry_category'].value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [38]:
NewCustomerList['wealth_segment'].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [39]:
NewCustomerList['state'].value_counts()

NSW    506
VIC    266
QLD    228
Name: state, dtype: int64

In [40]:
NewCustomerList['owns_car'].value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [41]:
NewCustomerList['deceased_indicator'].value_counts()

N    1000
Name: deceased_indicator, dtype: int64

# **6) EXPLORING CUSTOMER DEMOGRAPHIC DATASET**

In [42]:
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [43]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [44]:
CustomerDemographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

In [45]:
#Checking for null values
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

There are missing values in 5 columns.

## Checking Completeness of Datasets

In [73]:
# looking for the null values
total_null_values = CustomerDemographic.isnull().sum()

# calculating total values
total_values = CustomerDemographic.count().sort_values(ascending=True)

# calculating the percentage of null values
null_values_percentage = total_null_values/total_values *100

# converting to dataframe of missing values
missing_values_CustomerDemographic = pd.concat({'Total Values' : total_values, 'Null_values': total_null_values, 'Percentage of Missing Values': null_values_percentage}, axis=1)

# display missing values
print(missing_values_CustomerDemographic)

                                     Total Values  Null_values  \
job_industry_category                        3344          656   
job_title                                    3494          506   
last_name                                    3875          125   
DOB                                          3913           87   
tenure                                       3913           87   
customer_id                                  4000            0   
first_name                                   4000            0   
gender                                       4000            0   
past_3_years_bike_related_purchases          4000            0   
wealth_segment                               4000            0   
deceased_indicator                           4000            0   
owns_car                                     4000            0   

                                     Percentage of Missing Values  
job_industry_category                                   19.617225  
job_t

In [46]:
#Checking for duplicate data
CustomerDemographic.duplicated().sum()

0

There are no duplicate values.

In [47]:
#Checking for uniqueness of each column
CustomerDemographic.nunique()

customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

# **6.1) Exploring the columns**

In [48]:
CustomerDemographic.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

In [49]:
CustomerDemographic['gender'].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

Certain categories are not correctly titled. The names in these categories are re-named.

In [50]:
#Re-naming the categories
CustomerDemographic['gender'] = CustomerDemographic['gender'].replace('F','Female').replace('M','Male').replace('Femal','Female').replace('U','Unspecified')
CustomerDemographic['gender'].value_counts()

Female         2039
Male           1873
Unspecified      88
Name: gender, dtype: int64

In [51]:
CustomerDemographic['past_3_years_bike_related_purchases'].value_counts()

16    56
19    56
67    54
20    54
2     50
      ..
8     28
95    27
85    27
86    27
92    24
Name: past_3_years_bike_related_purchases, Length: 100, dtype: int64

In [52]:
CustomerDemographic['DOB'].value_counts()

1978-01-30    7
1964-07-08    4
1962-12-17    4
1978-08-19    4
1977-05-13    4
             ..
1989-06-16    1
1998-09-30    1
1985-03-11    1
1989-10-23    1
1991-11-05    1
Name: DOB, Length: 3448, dtype: int64

In [53]:
CustomerDemographic['job_title'].value_counts()

Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: job_title, Length: 195, dtype: int64

In [54]:
CustomerDemographic['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [55]:
CustomerDemographic['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [56]:
CustomerDemographic['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [57]:
CustomerDemographic['default'].value_counts()

100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

In [58]:
CustomerDemographic = CustomerDemographic.drop('default', axis=1)

The values are inconsistent, hence dropping the column.

In [59]:
CustomerDemographic.head(5)

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp  Female   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  tenure  
0                Health      Mass Customer                  N      Yes    11.0  
1    Financial Services      Mass Customer                  N      Yes    16.0  
2              Property      Mass Customer                  N      Yes    15.0  
3                    IT      Mass Customer                  N       No     7.0  
4                   NaN  Affluent Customer                  N      Yes     8.0

In [60]:
CustomerDemographic['owns_car'].value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

In [61]:
CustomerDemographic['tenure'].value_counts()

7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
9.0     200
14.0    200
6.0     192
13.0    191
4.0     191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: tenure, dtype: int64

# **7) EXPLORING CUSTOMER ADDRESS DATASET**

In [62]:
CustomerAddress.head(5)

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [63]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [64]:
CustomerAddress.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

In [65]:
#Checking for null values.
CustomerAddress.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

There are no null values.

## Checking completeness of datasets

In [74]:
# looking for the null values
total_null_values = CustomerAddress.isnull().sum()

# calculating total values
total_values = CustomerAddress.count().sort_values(ascending=True)

# calculating the percentage of null values
null_values_percentage = total_null_values/total_values *100

# converting to dataframe of missing values
missing_values_CustomerAddress = pd.concat({'Total Values' : total_values, 'Null_values': total_null_values, 'Percentage of Missing Values': null_values_percentage}, axis=1)

# display missing values
print(missing_values_CustomerAddress)

                    Total Values  Null_values  Percentage of Missing Values
customer_id                 3999            0                           0.0
address                     3999            0                           0.0
postcode                    3999            0                           0.0
state                       3999            0                           0.0
country                     3999            0                           0.0
property_valuation          3999            0                           0.0


In [66]:
#Checking for duplicate values
CustomerAddress.duplicated().sum()

0

There are no duplicate values.

In [67]:
#Checking for uniqueness of each column
CustomerAddress.nunique()

customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

# **7.1) Exploring the columns of Customer Address**

In [68]:
CustomerAddress['postcode'].value_counts()

2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: postcode, Length: 873, dtype: int64

In [69]:
CustomerAddress['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [70]:
CustomerAddress['country'].value_counts()

Australia    3999
Name: country, dtype: int64

In [71]:
CustomerAddress['property_valuation'].value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

## Highlights of Consistency and Validity in NewCustomerList, Customer Demographic and Customer Address
NewCustomerList dataset had 1000 records with 23 columns, yet Customer Demographics have 4000 records with 13 columns and remaining in Customer Address with 6 columns using **customer_id** has key.
- Structure format of NewCustomerList must be consistent with Customer Demographic and Customer Address.
- There is no **customer_id** in NewCustomerList.
- Number of columns are inconsistent because in NewCustomerList there are **4 columns which are Unnamed** and they contain some values as well, however are not labeled so cannot be identified. Therefore, these 4 columns were dropped.
- There is one column in *NewCustomerList* which is **Value**, it is captured in a column of **float64** datatype but this was not captured before and is not present in *CustomerDemographic* or *CustomerAddress*.
- There is one column named **default** in *CustomerDemographic*, it is captured in a column of **object** datatype, some values are observed to be date values but this was not captured after and is not present in *NewCustomerList*.
- From remaining columns 5 columns are of datatype **int64** which are past_3_years_bike_related_purchases, tenure, postcode, property_valuation, and Rank.                                .
- DOB is the date column **datetime64** in format **YYYY-MM-DD**. The date format used to capture transaction date in Transactions is **MM/DD/YYYY**. It would be better if it is kept consistent.
- Rest of the columns are in **object** data type values but, deceased_indicator must have contain **boolean** like True and False.
- Data Captured in Gender column in the dataset CustomerDemographic is not consistent. It should be "Male", "Female" and "U" as per the NewCustomerList. Therefore, renaming was performed above.

## Highlights of Completeness in NewCustomerList, Customer Demographic and Customer Address
- In NewCustomerList 19.76% of job_industry_category values are missing almost similar to CustomerDemographic which is 19.61%.
- 11.85% of job_title values are missing in NewCustomerList a little less as compared to CustomerDemographic that has 14.48% of missing values.
- 3.22% of last_name values were missing in CustomerDemographic yet 2.98% of last_name values are missing in NewCustomerList.
- CustomerDemographic has 2.22% of missing DOB values which is slighlty decreased to 1.72% NewCustomerList.
- There is a 2.22% of missing tenure values in CustomerDemographic but there is no missing values of tenure in NewCustomerList.


## Checking Accuracy of Dataset

In [75]:
CustomerDemographic['DOB']

0      1953-10-12
1      1980-12-16
2      1954-01-20
3      1961-10-03
4      1977-05-13
          ...    
3995   1975-08-09
3996   2001-07-13
3997          NaT
3998   1973-10-24
3999   1991-11-05
Name: DOB, Length: 4000, dtype: datetime64[ns]

## Highlights of Accuracy in NewCustomerList, Customer Demographic and Customer Address
One date value is wrong. 1843 year is not possible.

## Checking Uniqueness of all the tables of the Dataset

In [76]:
# looking for duplicated values
duplicated_values = NewCustomerList.duplicated()

# number of duplicated values in dataset
print("The number of duplicated records in NewCustomerList dataset is {}".format(duplicated_values.sum()))

The number of duplicated records in NewCustomerList dataset is 0


In [77]:
# looking for duplicated values
duplicated_values = CustomerDemographic.duplicated()

# number of duplicated values in dataset
print("The number of duplicated records in CustomerDemographic dataset is {}".format(duplicated_values.sum()))

The number of duplicated records in CustomerDemographic dataset is 0


In [78]:
# looking for duplicated values
duplicated_values = CustomerAddress.duplicated()

# number of duplicated values in dataset
print("The number of duplicated records in CustomerAddress dataset is {}".format(duplicated_values.sum()))

The number of duplicated records in CustomerAddress dataset is 0


## Highlights of Uniqueness
All records are unique.

# **8) MERGE ALL CUSTOMER INFORMATION FROM THREE TABLES INTO ONE TABLE**

In [79]:
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  tenure              address  \
0      Mass Customer                  N      Yes      14     45 Shopko Center   
1      Mass Customer                  N       No      16    14 Mccormick Park   
2  Affluent Customer                  N       No      10  5 Colorado Crossing   
3  Affluent Customer                  N      Yes       5   207 Annamark Plaza   
4  Affluent Customer                  N       No      19    115 Montana Place   

   postcode state    country  property_valuation  Rank     Value  
0      4500   QLD  Australia                   6     1  1.718750  
1      2113   NSW  Australia                  11     1  1.718750  
2      3505   VIC  Australia                   5     1  1.718750  
3      4814   QLD  Australia                   1     4  1.703125  
4      2093   NSW  Australia                   9     4  1.703125

In [80]:
CustomerDemographic['customer_id'].iloc[-1]

4000

We need to add 'Customer Id' column to new_customer_lists to enable us merge the tables vertically

In [81]:
NewCustomerList.insert(0, 'customer_id', range(4001, 4001 + len(NewCustomerList)))
NewCustomerList.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0         Manufacturing      Mass Customer                  N      Yes   
1              Property      Mass Customer                  N       No   
2    Financial Services  Affluent Customer                  N       No   
3         Manufacturing  Affluent Customer                  N      Yes   
4    Financial Services  Affluent Customer                  N       No   

   tenure              address  postcode state    country  property_valuation  \
0      14     45 Shopko Center      4500   QLD  Australia                   6   
1      16    14 Mccormick Park      2113   NSW  Australia                  11   
2      10  5 Colorado Crossing      3505   VIC  Australia                   5   
3       5   207 Annamark Plaza      4814   QLD  Australia                   1   
4      19    115 Montana Place      2093   NSW  Australia                   9   

   Rank     Value  
0     1  1.718750  
1     1  1.718750  
2     1  1.718750  
3     4  1.703125  
4     4  1.703125

In [82]:
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp  Female   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  tenure  
0                Health      Mass Customer                  N      Yes    11.0  
1    Financial Services      Mass Customer                  N      Yes    16.0  
2              Property      Mass Customer                  N      Yes    15.0  
3                    IT      Mass Customer                  N       No     7.0  
4                   NaN  Affluent Customer                  N      Yes     8.0

We need to merge the Customer Demographic with the Customer Address table before joining with New Customer List

In [83]:
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [84]:
# Merge dataframes using the customer_id column
CustomerDemographic = pd.merge(CustomerDemographic, CustomerAddress, how='left', on='customer_id')
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp  Female   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0                Health      Mass Customer                  N      Yes   
1    Financial Services      Mass Customer                  N      Yes   
2              Property      Mass Customer                  N      Yes   
3                    IT      Mass Customer                  N       No   
4                   NaN  Affluent Customer                  N      Yes   

   tenure              address  postcode            state    country  \
0    11.0   060 Morning Avenue    2016.0  New South Wales  Australia   
1    16.0  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2    15.0                  NaN       NaN              NaN        NaN   
3     7.0   0 Holy Cross Court    4211.0              QLD  Australia   
4     8.0  17979 Del Mar Point    2448.0  New South Wales  Australia   

   property_valuation  
0                10.0  
1                10.0  
2                 NaN  
3                 9.0  
4                 4.0

In [85]:
NewCustomerList.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0         Manufacturing      Mass Customer                  N      Yes   
1              Property      Mass Customer                  N       No   
2    Financial Services  Affluent Customer                  N       No   
3         Manufacturing  Affluent Customer                  N      Yes   
4    Financial Services  Affluent Customer                  N       No   

   tenure              address  postcode state    country  property_valuation  \
0      14     45 Shopko Center      4500   QLD  Australia                   6   
1      16    14 Mccormick Park      2113   NSW  Australia                  11   
2      10  5 Colorado Crossing      3505   VIC  Australia                   5   
3       5   207 Annamark Plaza      4814   QLD  Australia                   1   
4      19    115 Montana Place      2093   NSW  Australia                   9   

   Rank     Value  
0     1  1.718750  
1     1  1.718750  
2     1  1.718750  
3     4  1.703125  
4     4  1.703125

In [86]:
new_df = pd.concat([CustomerDemographic, NewCustomerList], ignore_index=True, sort=False)
new_df

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp  Female   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
4995         4996       Ferdinand  Romanetti    Male   
4996         4997            Burk    Wortley    Male   
4997         4998        Melloney      Temby  Female   
4998         4999          Dickie    Cubbini    Male   
4999         5000           Sylas    Duffill    Male   

      past_3_years_bike_related_purchases        DOB  \
0                                      93 1953-10-12   
1                                      81 1980-12-16   
2                                      61 1954-01-20   
3                                      33 1961-10-03   
4                                      56 1977-05-13   
...                                   ...        ...   
4995                                   60 1959-10-07   
4996                                   22 2001-10-17   
4997                                   17 1954-10-05   
4998                                   30 1952-12-17   
4999                                   56 1955-10-02   

                         job_title job_industry_category     wealth_segment  \
0              Executive Secretary                Health      Mass Customer   
1           Administrative Officer    Financial Services      Mass Customer   
2               Recruiting Manager              Property      Mass Customer   
3                              NaN                    IT      Mass Customer   
4                    Senior Editor                   NaN  Affluent Customer   
...                            ...                   ...                ...   
4995                     Paralegal    Financial Services  Affluent Customer   
4996        Senior Sales Associate                Health      Mass Customer   
4997  Budget/Accounting Analyst IV    Financial Services  Affluent Customer   
4998             Financial Advisor    Financial Services      Mass Customer   
4999           Staff Accountant IV              Property      Mass Customer   

     deceased_indicator owns_car  tenure                   address  postcode  \
0                     N      Yes    11.0        060 Morning Avenue    2016.0   
1                     N      Yes    16.0       6 Meadow Vale Court    2153.0   
2                     N      Yes    15.0                       NaN       NaN   
3                     N       No     7.0        0 Holy Cross Court    4211.0   
4                     N      Yes     8.0       17979 Del Mar Point    2448.0   
...                 ...      ...     ...                       ...       ...   
4995                  N       No     9.0               2 Sloan Way    2200.0   
4996                  N       No     6.0         04 Union Crossing    2196.0   
4997                  N      Yes    15.0  33475 Fair Oaks Junction    4702.0   
4998                  N      Yes    19.0        57666 Victoria Way    4215.0   
4999                  N      Yes    14.0        21875 Grover Drive    2010.0   

                state    country  property_valuation    Rank  Value  
0     New South Wales  Australia                10.0     NaN    NaN  
1     New South Wales  Australia                10.0     NaN    NaN  
2                 NaN        NaN                 NaN     NaN    NaN  
3                 QLD  Australia                 9.0     NaN    NaN  
4     New South Wales  Australia                 4.0     NaN    NaN  
...               ...        ...                 ...     ...    ...  
4995              NSW  Australia                 7.0   996.0  0.374  
4996              NSW  Australia                10.0   997.0  0.357  
4997              QLD  Australia                 2.0   997.0  0.357  
4998              QLD  Australia                 2

Here, new_df is not merged with transactions because in each customer id there would many transactions which causes redundancy therefore uniquess will be lost.

## **Sample code is given below in case if we plan to merge the new_df with the transactions**

In [87]:
# We would assign the new dataframe to customer demographic where we will use one dataframe to capture all the customer information
# CustomerDemographic = new_df
# Merge dataframes using the customer_id column
# CustomerDemographic = pd.merge(CustomerDemographic, Transactions, how='left', on='customer_id')
# CustomerDemographic.head()